src/conq/navigation_lib/openai_nav/openai_nav_client.py

```diff
@@ -0,0 +1,96 @@
+
+# Change this based on which API you are using
+os.environ["OPENAI_API_KEY"] = OPENAI_API_KEY.get()
+
+ONE_MINUTE = 60  # Constant for seconds in a minute
+
+
+class OpenAINavClient:
+    def __init__(self, locations: List[str]):
+        # Sets the current working directory to be the same as the file.
+        os.chdir(os.path.dirname(os.path.abspath(__file__)))
+
+        try:
+            if load_dotenv(".env") is False:
+                raise TypeError
+        except TypeError:
+            print("Unable to load .env file.")
+            quit()
+
+        # Create Azure client
+        # self.llm_client = AzureOpenAI(
+        #     api_key=os.environ["OPENAI_API_KEY"],
+        #     api_version=os.environ["API_VERSION"],
+        #     azure_endpoint=os.environ["openai_api_base"],
+        #     organization=os.environ["OPENAI_organization"],
+        # )
+        self.llm_client = OpenAI(
+            api_key=os.environ["OPENAI_API_KEY"],
+        )
+
+        # Locations in our environment
+        self.locations = locations
+
+        # System context to give to the LLM before asking question
+        self.system_context = "Assistant is a large language model trained by OpenAI."
+
+    @limits(calls=3, period=ONE_MINUTE)
+    def find_probable_location_for_object(self, object: str):
+        # Build the query string
+        query = "I observe the following structures while exploring a small-scale farm:"
+
+        for index in range(len(self.locations)):
+            query += f"\n{index + 1}. {self.locations[index]}"
+
+        query += f"\n\nPlease rank the structure based on how likely I am to find a {object} in them. Please provide the response in a plain text string representing a python dictionary, {{\"name\": probability}} and only include the location with whitespace connected with _ as the keys and probabilities as the values"
+        # Make the query
+        response = self.llm_client.chat.completions.create(
+            messages=[
+                {"role": "system", "content": self.system_context},
+                {"role": "user", "content": query},
+            ],
+            model=os.environ["model"],
+            temperature=0,
+        )
+
+        try:
+            probabilities_out = ast.literal_eval(response.choices[0].message.content)
+        except:
+            # Try to send the bad output back through chatGPT
+            num_tries = 0
+            for i in range(3):
+                num_tries += 1
+                query = f"This output is not in a plain text string representing a dictionary, please make it formatted correctly: {response.choices[0].message.content}"
+                response = self.llm_client.chat.completions.create(
+                    model=os.environ["model"],
+                    messages=[
+                        {"role": "system", "content": self.system_context},
+                        {"role": "user", "content": query},
+                    ],
+                    temperature=0,
```

Comment:
As I understand it, `temperature` controls the degree of "randomness" in the response to some extent, right? As in, the higher the temperature, the more likely ChatGPT is to select the next word that is not the most likely, right? Or do I have it wrong?

I'm asking this as I'm curious if a temperature of 0 is the default setting. If it isn't, could you add a comment that explains why you chose `temperature=0`?

Reply:
Yes exactly! It defaults to 1 and can be anything 1-2. I chose 0 because it was getting too verbose with the responses, even though the dictionary was correct. For example, it would always say something like "Here is the dictionary that you wanted: " even when I would instruct it to not include explanation. I will add a quick comment on this

Reply:
Great! Thanks for adding the explanation